Finding unimportant features

omidkj opened the pull request:
implementing enhancement issue #10 

in the current state of this PR it's finding missing values, additional work to be done

Resolves #10

omidkj committed:
Finding unimportant features: missing values

- This method receives a ndarray with a specified threshold (between 0 and 1)
  for missing values

omidkj committed:
Finding unimportant features: missing values

- This method receives a ndarray with a specified threshold (between 0 and 1)
  for missing values

omidkj committed:
Finding unimportant features: single unique value

- This method finds any features that have only a single value in their columns.

omidkj committed:
Adding more information about columns with missing values and single values.

- Print column number and the missing value fraction of that column.
- Print the column number and the single value of that column.

omidkj committed:
Finding highly correlated features:

-This method receives a ndarray with a specified threshold for the correlation coefficient value as parameters. It uses a Pearson correlation method and returns features that have correlation coefficient value more than specified value.

omidkj committed:
Find low important features based on Random forest importance result:

- This method receives a ndarray with a specified threshold for the normalized importance value (between 0 and 1). For instance a threshold of 0.45 means find feature with that their normalized importance value is below 0.45.

omidkj committed:
Adding comments and notes to the code.

- notes to reference the code in future
- Describing the modules and functions

omidkj committed:
Adding type hints to the code.

omidkj commented:
rebased against current master.

NikitaThomas commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,188 @@
+                low_importance +=[low_imp[i]]
+
+        self.low_importance_col = np.array(low_importance)
+        return self.low_importance_col
```
Hi Omid, I reviewed your module and thought it was very well documented because of your clear and concise docstrings, which made reviewing your code way more efficient. Overall I believe this will be a very useful module!  Do you have a specific reason why the default threshold is 40% ?

rohitchadaram replied:
Great work Omid, it really is self explanatory the way you added all the comments : function definitions which are usually missed and overlooked. I really like the code structure. But I had the same question in mind as Nikita pointed out as well, How could you ever determine for a dataset which is a good threshold(40%,50%?) for missing or even correlation threshold(0.9) there could be cases when you have correlation value as 0.8 which is still a high correlation value and can say the features are colinear. Also if your dataset is indeed small then finding columns with greater than 40% missing values to eliminate/ignore them might not make sense. How about considering the size of the dataset for figuring out a threshold value ?

omidkj replied:
Thank you @NikitaThomas :1st_place_medal: for reviewing my code and for your kind comment. To be honest, I don't know how to come up with a reasonable default value! I'll do more research and update my code. Meanwhile, if you have any suggestion please send them my way.

omidkj replied:
Thanks @rohitchadaram :1st_place_medal:  for taking time and reviewing my code. As we discussed int the class it might need more research to find a reasonable default value for a parameter. I really like your idea of considering the size of the dataset as a driving factor to find the threshold. I'll try to find a way to incorporate that in my code.

omidkj commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,188 @@
+
+        """
+        # Performing the train_test_split
+        X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.25)
```
I think here, I might need to assume that the data has been already split into training and test. Therefore, this method is not responsible of the leakage of data from test to train set.

omidkj replied:
Instead, I think I might need to add K-fold cross-validation.

omidkj committed:
adding Kfold cross validation

- removing train_test_split
- adding if __name__ == "__main__" to include some examples of creating an object from class and calling methods.

rohitchadaram commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+
+
+if __name__ == "__main__":
+    data_np = np.genfromtxt('data.csv', delimiter=',', skip_header=1, missing_values='None', filling_values=np.nan)
```
Hi Omid, I definitely do not have any comments regarding code quality. I think you done a great job. Let me nitpick few things which you could improve open : I noticed you had multiple for loops embedded in the code for correlation, I wonder if there is any way to avoid that as you see it basically adds a lot of run-time complexity to your code when you have a relatively big dataset. You have two inner for loops which to think there is a way to avoid repetition using zip in python, I'm not entirely sure if it's feasible but you can try. Also you are loading the hard-coded name of the dataset which I earlier did too, so a suggestion to load the filepath name through command line argument like I did to just avoid any hard-coding of parameters which you have done efficiently throughout the rest of your code.

EHWUSF replied:
I'll guess that this section is set up for test purposes, especially since you have imposed constraints on the number of attributes in the data file, but whatever the purpose, it should be explained with some comments.  If it is for some test purposes, then the applicable test file should also be provided, and the expected results mentioned in the comments and/or directly checked.

I recall there having been some discussion amongst the class of possibly standardizing on a particular data file for test purposes, and if that's what you were expecting here then it should at least be suitably documented and ideally someone would actually submit a PR with that datafile so that it could be merged prior to testing out all the various PRs dependent on it.

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+    X_train, X_test, y_train, y_test = train_test_split(features_np, target_np, test_size=0.25)
+    frs = FeatureRemover()
+    correlated_indices = frs.colinear(X_train, corr_thresh=0.7)
+    low_importance_indices = frs.low_importance(X_train, y_train, importance_thresh=0.12)
```
trying to run this with a small test data set, I get the following exception (which was a bit extra puzzling since my dataset had 13 samples but the message refers to 9 samples). 
```
  File "/home/willgohs/PycharmProjects/HS628/HS68_2018_Project_1/hs628_tools/feature_missing.py", line 201, in <module>
    low_importance_indices = frs.low_importance(X_train, y_train, importance_thresh=0.12)
  File "/home/willgohs/PycharmProjects/HS628/HS68_2018_Project_1/hs628_tools/feature_missing.py", line 162, in low_importance
    for train_index, test_index in skf.split(features, target):
  File "/usr/local/lib/python2.7/dist-packages/sklearn/model_selection/_split.py", line 320, in split
    n_samples))
ValueError: Cannot have number of splits n_splits=10 greater than the number of samples: 9.
```
 It's not particularly surprising that it would fail with a small dataset, but applicable constraints should be mentioned in a docstring.   And  when I tried making my test file modestly bigger, the error message shifted to
```
  File "/usr/local/lib/python2.7/dist-packages/sklearn/model_selection/_split.py", line 575, in _make_test_folds
    % (self.n_splits))
ValueError: All the n_groups for individual classes are less than n_splits=10
```

Particularly since you don't give the user the ability to control n_splits, it would be best to do some safety checking and/or exception handling.  I'll make a note to review the use of ```try: except``` clauses in class.

omidkj replied:
fixing this by setting the n_splits as a parameter n_folds with a default value of 'None', then set the n_folds based on the dataset size inside the method: 

```python
    def low_importance(self, features, target, target_type="classification", importance_thresh=None, n_folds=None):
         if importance_thresh is None:
             importance_thresh = 1/features.shape[1]

        if n_folds is None:
            if (features.shape[0] < 100) and (features.shape[0] > 20):
                n_folds = 5
            elif features.shape[0] < 21 and (features.shape[0] > 3):
                n_folds = 3
            elif features.shape[0] < 4:
                n_folds = 1
            else:
                n_folds = 10
```

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+        for i in range(data.shape[1]):
+            # finding the unique value of each column and if the lenghth of that equals 1 means
+            # that feature has only one sibgle value, then save the index of that column in single_unique
+            if len(np.unique(data[:, i], axis=0)) == 1:
```
Pycharm flags this line, for np.unique not having an ```axis``` parameter, and I get a corresponding exception when I try using the ```single_value``` method, both with numpy 1.14.5 and with numpy 1.15 (which I believe is the most current version).

Does this code run for you?  What version of python and numpy are you using?

omidkj replied:
@EHWUSF I'm using python 3.6.6 with numpy 1.14.5 and don't see this flag.

omidkj replied:
I'll remove the axis since it doesn't make any difference and it might be flagged in other versions.

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+        1. which_missing: Returns columns with a missing percentage greater than a specified threshold 2.
+        single_value: Returns columns with a single unique value 3. colinear: Returns collinear columns with a
+        correlation greater than a specified correlation threshold 4. low_importance: Returns features that are
+        ranked low importance in the Random Forest Regressor or Classifier #Notes: using the idea adapted from:
```
Adapting ideas is generally fine (subject to the possibility of any applicable patents) but generally doesn't result in so much code similarity; adapting someone's code is also often fine (depending on, among other things, what license they grant for their code) but should be clearly acknowledged as such ("adapted from X" vs "using ideas from X") and the commits can be used to clearly show which part is old/borrowed versus which part is new.  (i.e. a starting point of old code followed by a series of adaptation commits)

omidkj replied:
@EHWUSF My code is entirely different from the source I mentioned here since I'm using numpy arrays instead of Pandas. 
Would you recommend changing the language from "using the idea adapted from X" to "adapted from X" ?

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+        # type: (ndarray, float) -> ndarray
+
+        """
+        Find columns with a fraction of missing data more than pre-dedfined threshold
```
Should also mention that information is printed out.

omidkj replied:
Thanks, I'll add that to the docstring

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+    X_train, X_test, y_train, y_test = train_test_split(features_np, target_np, test_size=0.25)
+    frs = FeatureRemover()
+    correlated_indices = frs.colinear(X_train, corr_thresh=0.7)
+    low_importance_indices = frs.low_importance(X_train, y_train, importance_thresh=0.12)
```
Why not testing the other two methods?

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+        # type: (ndarray, float) -> ndarray
+
+        """
+        Find columns with correlation coefficient value more than specified value
```
should mention the information printed out

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+                if abs(corr_matrix[i, j]) > corr_thresh:
+                    corr_pairs += [(i, j)]
+
+        print('%d features with a correlation greater than %0.2f:\n\n' % (len(corr_pairs), corr_thresh))
```
Although the detail lines are more clear, this summary line is a bit misleading--  isn't really the number of features with correlation greater than X, it is the number of correlated pairs-- with a group of 4 features all correlated above the threshold, it reports "6 features with a correlation greater" not "4 features with a correlation greater".

omidkj replied:
Fixing this with:

```python
        print('%d features with a correlation greater than %0.2f:\n\n' % (len(np.unique(corr_pairs)), corr_thresh))
```

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+        for i in range(train_data.shape[1]):
+            for j in range(train_data.shape[1]):
+                # Calculating the corrcoef of each pair and save in corr_matrix
+                corr_matrix[i, j] = np.corrcoef(train_data[:, i], train_data[:, j])[0, 1]
```
This produces a symmetric matrix, which is why you only actually need to check values below where j > i.  Since this calculation is relatively expensive and you don't appear to try looking up any values for j <= i, there is no point using up cpu to calculate them (and even if you wanted to be able to refer to them that way, you'd still only need to do half of these calculations and just set both entries).

And if you look more closely at ```np.corrcoef()```, I think you'll see you could (probably more efficiently) determine corr_matrix without these nested loops at all.

omidkj replied:
removing the first nested loop and adding this line instead to generate the correlation matrix:

```python
corr_matrix = np.corrcoef(train_data, rowvar=False)
```

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+                target:     numpy ndarrays, shape = (1, ) target can be binary if target_type is classification or
+                            continuous if target_type is regression
+
+                target_type : string defines the RandomForest method if target is continuous or binary. default value
```
I think this comment should mention both possible values for the parameter, not just one, and be a bit more clearly phrased in general.

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,201 @@
+        # type: (ndarray, ndarray, str, float) -> ndarray
+
+        """
+        Find columns with their normalized importance value below pre-defined threshold using RandomForestRegressor or
```
docstring should mention there is output

omidkj committed:
Applying some changes based on prof. Eric's comments:

- Removing the class structure and adding functions instead of methods
- Changing the wording of the code adoption
- Adding function outputs to the docstrings
- Removing the first nested loop inside the Collinear function, using np.corrcoef() instead
- Setting the n_splits as a parameter n_folds with a default value of 'None', then set the n_folds based on the dataset size inside the method:
- Removing the main function that was created for test purposes.

omidkj committed:
changing the wording of the code adoption

EHWUSF commented on hs628_tools/feature_missing.py:
```diff
@@ -0,0 +1,200 @@
+            if abs(corr_matrix[i, j]) > corr_thresh:
+                corr_pairs += [(i, j)]
+
+    print('%d features with a correlation greater than %0.2f:\n\n' % (len(np.unique(corr_pairs)), corr_thresh))
```
I think this change doesn't actually change the issue & output at all, because all the pairs were already unique.  I think that the simplest reasonable fix is to *instead simply change the text* to read something like ```'%d feature pairs with...'```

Something to consider for the future that I think would be even better but a bit trickier (and not something I'd expect you to tackle now), would be to report strongly correlated groups of features.   Suppose that features a, b, c, and d were all strongly correlated to each other, and that feature e was strongly correlated to feature a but not to the other features.  The current approach would report 7 pairs of strongly correlated features, but how does the user then decide that to do with that?  If you consider the features to be vertices in a graph, with edges between those features that had pair-wise correlation above the threshold, and determine [the cliques of the graph](https://en.wikipedia.org/wiki/Clique_(graph_theory)), then the strongly correlated groups would be the maximal cliques (you'd only report those with size larger than 1 vertex), and a reasonable feature reduction approach would be to have a single proxy feature for each maximal clique (or perhaps not each, just enough to make a clique cover).   In this example with features a-e, rather than getting a listing of 7 feature pairs, the user would be told
```
With correlation threshold X%,
Features a, b, c, d form  a strongly correlated group
Features a, e form a strongly correlated group
```

omidkj committed:
Fixing the issue in collinear function by changing the test to feature pairs with

omidkj committed:
Adding a plot to which_missing function to plot all the columns with their missing data fraction against the threshold line.

omidkj committed:
Setting the default value of missing threshold based on the size of data based on Rohit's suggestion

omidkj committed:
Adding a plot for importance values in low_importance function

omidkj committed:
Update README.md
